### PR TITLE
Add OLMo2/3 models support in fairseq2

### LINKED ISSUE #1402: Implement OLMO2/3 models support in Fairseq2
**Is your feature request related to a problem? Please describe:**
A clear and concise description of what the problem is.

**Describe the solution you would like:**

**Describe the alternatives you have considered:**
A clear and concise description of any alternative solutions or features you have considered.

**Additional Context:**
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by YunchaoYang**
**What does this PR do? Please describe:**

Add OLMo2 and OLMo3 model architecture support in fairseq2.

- **OLMo2**: 1B, 7B, 13B, 32B
- **OLMo3**: 7B, 32B

Both architectures share a unified `olmo` module. The key architecture features:

- **`OLMORMSNorm`**: The order of operations is normalize → multiply by weight → cast to original dtype (differs from standard RMSNorm which casts before multiplying).

- **`OLMOTransformerLMDecoderLayer`**: Uses a custom Post-Norm order: Attention/FFN → Norm → Add Residual. This differs from both standard Pre-Norm (Norm → Attention/FFN → Add Residual) and standard Post-Norm (Attention/FFN → Add Residual → Norm).

- **`OLMOMultiheadAttention`**: Inherits directly from the `MultiheadAttention` abstract base and inlines Q/K/V projection setup.
  - Q/K Norm applied in the order: Project → Normalize → Reshape → RoPE.
  - Supports both MHA and GQA (OLMo2-32B and OLMo3-32B use GQA; all other sizes use MHA).
  - KV-caching for incremental (auto-regressive) decoding.

- **Rotary Encoding**:
  - OLMo2 uses `ReferenceRotaryEncoder` (standard RoPE) directly.
  - OLMo3 uses a standalone `YaRNRotaryEncoder` for long-context extension (8K → 65K), implementing frequency-dependent scaling and attention scaling (mscale).

- **OLMo3 Hybrid Attention** (sliding window + full attention):
  - Most layers use sliding window attention (window=4096); every 4th layer and the final layer use full global attention.
  - Per-layer RoPE: sliding window layers use standard `ReferenceRotaryEncoder`, full attention layers use `YaRNRotaryEncoder`.

**Testing:**

- Integration tests verify output consistency with HuggingFace Transformers for OLMo2-1B and OLMo3-7B.
- Unit tests cover attention (Q/K norm order), YaRN RoPE, HF weight conversion, and incremental decoding.
- An SFT training recipe is included for OLMo2-1B on GSM8K.

Fixes #1402 

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [x] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**?
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/models/olmo/attention.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        output_proj_bias: bool | None = None,
+        state_factory: AttentionStateFactory | None = None,
+        gangs: Gangs | None = None,
+        device: Device | None = None,
+        dtype: DataType | None = None,
+    ) -> None:
+        super().__init__()
+
+        # --- num_heads / num_key_value_heads validation ---
+        self.num_heads = num_heads
```
Comment (cirquit) [inline src/fairseq2/models/olmo/attention.py]: Looks like KV-caching is missing, which means OLMo models cannot do auto-regressive generation - only full-sequence forward passes.

Check out StandardMultiheadAttention.forward() which has state_bag.maybe_get_state() for K/V caching. Either restore that logic or document that generation is unsupported and track as a follow-up.

File: src/fairseq2/models/olmo/config.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    dropout_p: float = 0.0
+    """The dropout probability on outputs of Transformer layers."""
+
+    # TODO check the init_std == initializer_range?
+    init_std: float | None = None
+    """
+    If not ``None``, the standard deviation to initialize input embeddings and
+    projection weights; otherwise, ``model_dim ** -0.5`` will be used instead.
+    """
+    # initializer_range: float = 0.02
```
Comment (cirquit) [inline src/fairseq2/models/olmo/config.py]: Unresolved TODO and dead code (`initializer_range`)

File: src/fairseq2/models/olmo/config.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+    mscale_all_dim: float = 0.0
+    """Dimension-wise scaling parameter for YaRN."""
+
+    truncate: bool = True
+    """If True, truncate correction range bounds to integers. Default: True."""
+
+
+@dataclass(kw_only=True)
+class OLMOConfig(LLaMAConfig):
```
Comment (cirquit) [inline src/fairseq2/models/olmo/config.py]: Is there a need to inherit from LLaMAConfig if you redefine almost every parameter? Maybe worth to keep solo.

File: src/fairseq2/models/olmo/attention.py (hunk 1 above)
Comment (cirquit) [inline src/fairseq2/models/olmo/attention.py:88]: Looks like KV-caching is missing, which means OLMo models cannot do auto-regressive generation - only full-sequence forward passes.

Check out StandardMultiheadAttention.forward() which has state_bag.maybe_get_state() for K/V caching. Either restore that logic or document that generation is unsupported and track as a follow-up.

File: src/fairseq2/models/olmo/config.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    """
+    The method to use to scale ``init_std`` per layer. If 'none', no scaling
+    will be applied. If 'layer', ``init_std`` will be scaled by the depth of
+    the layer. If 'stack', ``init_std`` will be scaled by the total depth of
+    the decoder.
+    """
+
+    # TODO check if it is used in olmo
+    # shard_embed_dim: bool = False
+    """If ``True``, shards the embedding dimension for tensor parallelism."""
```
Comment (cirquit) [inline src/fairseq2/models/olmo/config.py]: `shard_embed_dim` is commented out as `False` but inherited as `True` from `LLaMAConfig`. The factory uses config.shard_embed_dim at line 183, so OLMo models silently get `True`. Either uncomment and set to False, or confirm True is correct and remove the comment.

I think setting to `True` is sane here but you have more context.

File: src/fairseq2/models/olmo/interop.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    - No input_layernorm (Pre-Norm)
+    - Has post_attention_layernorm (after attention)
+    - Has post_feedforward_layernorm (after FFN)
+    - Has Q/K Norm in attention layers
+    """
+    # Handle legacy format (wrapped in "model" key)
+    try:
+        state_dict = cast(dict[str, object], state_dict["model"])
+    except KeyError:
+        pass
```
Comment (cirquit) [inline src/fairseq2/models/olmo/interop.py]: No need for this given that no olmo3 models were trained on fairseq2 v0.4.

File: src/fairseq2/models/olmo/interop.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    if "model.embed_tokens.weight" in state_dict:  # HuggingFace format
+        # Note: OLMO may need RoPE weight permutation similar to LLaMA
+        # This depends on the actual checkpoint format. For now, we'll check
+        # if weights need permutation by comparing shapes.
+
```
Comment (cirquit) [inline src/fairseq2/models/olmo/interop.py]: Is this still the case?

File: src/fairseq2/models/olmo/normalization.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""OLMO-specific normalization layers."""
+
+from __future__ import annotations
+
+from typing import Callable, Sequence
```
Comment (cirquit) [inline src/fairseq2/models/olmo/normalization.py]: import `Callable`, `Sequence` from `collections.abc`, not `typing`

File: src/fairseq2/models/olmo/tokenizer.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    HuggingFaceTokenModel,
+    load_hg_token_model,
+)
+from fairseq2.device import Device
+from fairseq2.error import NotSupportedError
+
+# Chat template with assistant mask support.
+
+## this article shows the hg implementation on
+# https://huggingface.co/allenai/OLMo-2-1124-7B-Instruct/blob/main/tokenizer_config.json
```
Comment (cirquit) [inline src/fairseq2/models/olmo/tokenizer.py]: Maybe move inside the class as docstring?

File: src/fairseq2/models/olmo/tokenizer.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+"""
+
+# chat template example
+"""
+# <|endoftext|><|user|>
+How are you doing?
+<|assistant|>
+I'm just a computer program, so I don't have feelings, but I'm
+functioning as expected. How can I assist you today?<|endoftext|>
+"""
```
Comment (cirquit) [inline src/fairseq2/models/olmo/tokenizer.py]: Loose docstring, maybe move inside the class as example?

File: tests/integration/models/test_olmo.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""Integration tests for all OLMO model variants."""
```
Comment (cirquit) [inline tests/integration/models/test_olmo.py]: So I'd propose to remove the integration test after I we ok this PR given the heavy-weight nature of it, we can't make framework tests dependable on local paths.

If fine with you, after everything else has been verified to work we can remove this test before merging.

File: src/fairseq2/composition/models.py [hunk 11]
```diff
@@ -308,6 +315,21 @@ def _register_model_families(container: DependencyContainer) -> None:
...
+        container,
+        OLMO_FAMILY,
+        kls=TransformerLM,
+        config_kls=OLMOConfig,
+        factory=create_olmo_model,
+        state_dict_converter=convert_olmo_state_dict,
+        compiler=compile_transformer_lm,
+        fsdp_applier=apply_fsdp_to_transformer_lm,
+        layerwise_ac_applier=apply_ac_to_transformer_lm,
+    )
```
Comment (cirquit) [inline src/fairseq2/composition/models.py:330]: Missing HuggingFaceConverter registration, LLaMA and Qwen both register one here! Without it, checkpoints saved during training won't export to HF format. Add a converter in interop.py
   (implement to_hg_config and to_hg_state_dict) and register it. See src/fairseq2/models/qwen/interop.py:51 for reference.

File: src/fairseq2/models/olmo/config.py (hunk 3 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/config.py]: Fixed. OLMOConfig is an independent config class now.

File: src/fairseq2/models/olmo/config.py (hunk 2 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/config.py]: Dead code removed

File: src/fairseq2/models/olmo/attention.py (hunk 1 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/attention.py:88]: Fixed. The kv caching logic in OLMO is added, uses state_bag.maybe_get_state() / set_state() with AttentionState and FullAttentionState, matching the self-attention branch in StandardMultiheadAttention.forward(). 
A dedicated test passed for auto-regressive generation.

File: src/fairseq2/models/olmo/config.py (hunk 4 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/config.py]: Fixed. The shard_embed_dim for OLMO is explicitly set to be True.

File: src/fairseq2/models/olmo/interop.py (hunk 5 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/interop.py]: Removed these deprecated code.

File: src/fairseq2/models/olmo/interop.py (hunk 6 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/interop.py]: The comment was removed. 

OLMo HuggingFace checkpoints don't require RoPE weight permutation (unlike LLaMA). The Q/K projection weights are stored in standard layout, so convert_olmo_state_dict only needs the regex key-map rename, no shape manipulation. Verified by the integration tests that compare HF vs fairseq2 outputs.

File: src/fairseq2/models/olmo/normalization.py (hunk 7 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/normalization.py]: Fixed. imports now come from collections.abc.

File: src/fairseq2/models/olmo/tokenizer.py (hunk 8 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/tokenizer.py]: The stale comments are removed. 
The chat template is kept as a module-level Final constant (OLMO_HG_CHAT_TEMPLATE)

File: src/fairseq2/composition/models.py (hunk 11 above)
Comment (YunchaoYang) [reply src/fairseq2/composition/models.py:330]: Done. 

_OLMOHuggingFaceConverter is registered at line 335 of [composition/models.py](https://github.com/facebookresearch/fairseq2/blob/b67f714ecabed99376ee6003a4ed43fc9e7b05d7/src/fairseq2/composition/models.py#L320-L336) (following the same pattern as LLaMA and Qwen .) and implements both to_hg_config (with OLMO2/OLMO3 branching) and to_hg_state_dict in [interop.py](https://github.com/facebookresearch/fairseq2/blob/b67f714ecabed99376ee6003a4ed43fc9e7b05d7/src/fairseq2/models/olmo/interop.py#L79).

File: tests/integration/models/test_olmo.py (hunk 10 above)
Comment (YunchaoYang) [reply tests/integration/models/test_olmo.py]: Agreed. The integration tests for OLMO are removed in commit 5f492f33.

File: src/fairseq2/models/olmo/tokenizer.py (hunk 9 above)
Comment (YunchaoYang) [reply src/fairseq2/models/olmo/tokenizer.py]: Fixed. The loose docstring example was removed. The chat template is now a clean module-level Final constant

File: src/fairseq2/models/olmo/attention.py [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+        seqs_layout: BatchLayout,
+        keys: Tensor,
+        keys_layout: BatchLayout,
+        values: Tensor,
+        bias_cache: AttentionBiasCache,
+        *,
+        state_bag: IncrementalStateBag | None = None,
+    ) -> Tensor:
+        """Forward pass with OLMO2 rotary encoding semantics."""
+        # Project Q, K, V with normalization
```
Comment (cirquit) [inline src/fairseq2/models/olmo/attention.py]:
```suggestion
```
Unnecessarily verbose comment

File: src/fairseq2/models/olmo/config.py [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+
+def register_olmo_configs(container: DependencyContainer) -> None:
+    """Register OLMO model configurations (OLMO2 and OLMO3)."""
+    arch = ConfigRegistrar(container, OLMOConfig)
+
+    # OLMO2 Model Configurations
+    @arch("olmo2_1b")
+    def olmo_2_1b() -> OLMOConfig:
+        """OLMO2 1B model configuration."""
+        # All parameters are already defaults in OLMOConfig
```
Comment (cirquit) [inline src/fairseq2/models/olmo/config.py]:
```suggestion
```
Verbose comment.

File: src/fairseq2/models/olmo/config.py [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+
+    If None, uses standard RoPE without scaling (default for OLMO2/3 base models).
+    """
+
+
+def register_olmo_configs(container: DependencyContainer) -> None:
+    """Register OLMO model configurations (OLMO2 and OLMO3)."""
+    arch = ConfigRegistrar(container, OLMOConfig)
+
+    # OLMO2 Model Configurations
```
Comment (cirquit) [inline src/fairseq2/models/olmo/config.py]:
```suggestion
```
Verbose comment.

File: src/fairseq2/models/olmo/config.py [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+
+    @arch("olmo2_32b")
+    def olmo_2_32b() -> OLMOConfig:
+        """OLMO2 32B model configuration with GQA.
+
+        Uses Grouped Query Attention (GQA) for efficiency.
+        """
+        config = OLMOConfig()
+
+        # Model dimensions for 32B
```
Comment (cirquit) [inline src/fairseq2/models/olmo/config.py]:
```suggestion
```
